pkg/operator/controller/gatewayclass/controller.go

```diff
@@ -0,0 +1,97 @@
+package gatewayclass
+
+import (
+	"context"
+
+	logf "github.com/openshift/cluster-ingress-operator/pkg/log"
+
+	"k8s.io/client-go/tools/record"
+
+	gatewayapiv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	controllerName = "gatewayclass_controller"
+
+	// OpenShiftGatewayClassControllerName is the string by which a
+	// gatewayclass identifies itself as belonging to OpenShift Istio.  If a
+	// gatewayclass's spec.controllerName field is set to this value, then
+	// the gatewayclass is ours.
+	OpenShiftGatewayClassControllerName = "openshift.io/gateway-controller"
+)
+
+var log = logf.Logger.WithName(controllerName)
+
+// NewUnmanaged creates and returns a controller that watches gatewayclasses and
+// installs and configures Istio.  This is an unmanaged controller, which means
+// that the manager does not start it.
+func NewUnmanaged(mgr manager.Manager, config Config) (controller.Controller, error) {
+	reconciler := &reconciler{
+		config:   config,
+		client:   mgr.GetClient(),
+		cache:    mgr.GetCache(),
+		recorder: mgr.GetEventRecorderFor(controllerName),
+	}
+	c, err := controller.NewUnmanaged(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+	isOurGatewayClass := predicate.NewPredicateFuncs(func(o client.Object) bool {
+		class := o.(*gatewayapiv1beta1.GatewayClass)
+		return class.Spec.ControllerName == OpenShiftGatewayClassControllerName
+	})
+	if err := c.Watch(&source.Kind{Type: &gatewayapiv1beta1.GatewayClass{}}, &handler.EnqueueRequestForObject{}, isOurGatewayClass); err != nil {
```

Comment:
Have we definitely decided that no other "custom" GatewayClasses can be created?

Reply:
I'm not sure what you mean.  We only care about gatewayclasses that specify our controller name.  The cluster admin absolutely can create gatewayclasses that specify other controller names, but then any such named controllers (not ours) are responsible for managing those gatewayclasses.